### PR TITLE
nixos/uboot-builder: allow setting specific dtb

### PR DESCRIPTION
**Pull request by c0deaddict**
###### Description of changes

Uboot on Raspberry Pi's is loading the wrong DTB file for some unknown reason. For example on a Pi 3B it's loading bcm2837 instead of bcm2710. This has been corrected with the following fixup on the `linux-rpi` kernel: https://github.com/NixOS/nixpkgs/blob/b235fa2abc957923e7381bfca84f54bed453c6e9/pkgs/os-specific/linux/kernel/linux-rpi.nix#L58

However, this fixup doesn't work anymore if custom `hardware.deviceTree.overlays` are used, since the overlays are compiled from source (?). I've looked for a proper workaround and came up with this patch. This patch allows specifying a specific DTB that needs to be loaded, instead of Uboot choosing one. The `boot.loader.generic-extlinux-compatible` [already supports this option](https://github.com/NixOS/nixpkgs/blob/7b38b03d76ab71bdc8dc325e3f6338d984cc35ca/nixos/modules/system/boot/loader/generic-extlinux-compatible/default.nix#L57) but this isn't passed by the raspberrypi uboot builder.

For the Pi 3B i've used this setting to fixate the correct DTB:
```
hardware.deviceTree.name = "broadcom/bcm2710-rpi-3-b.dtb";
```

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/system/boot/loader/raspberrypi/raspberrypi.nix [hunk 1]
```diff
@@ -4,13 +4,16 @@ with lib;
...
+  dtCfg = config.hardware.deviceTree;
 
   builderUboot = import ./uboot-builder.nix { inherit pkgs configTxt; inherit (cfg) version; };
   builderGeneric = import ./raspberrypi-builder.nix { inherit pkgs configTxt; };
 
   builder =
     if cfg.uboot.enable then
-      "${builderUboot} -g ${toString cfg.uboot.configurationLimit} -t ${timeoutStr} -c"
+      "${builderUboot} -g ${toString cfg.uboot.configurationLimit} -t ${timeoutStr}"
+        + lib.optionalString (dtCfg.name != null) " -n ${dtCfg.name}"
```
Comment (klemensn) [inline nixos/modules/system/boot/loader/raspberrypi/raspberrypi.nix:15]: `dtCfg.name` is now an arbitrary user-provided string, so best use `lib.strings.escapeShellArg`.

Better yet, always sanitze:
```
builder = escapeShellArgs (
  if cfg.uboot.enable then [
    builderUboot
    "-g" "${toString cfg.uboot.configurationLimit}"
    "-t" timeoutStr
  ] ++ lib.optionals (dtCfg.name != null) [
    "-c" dtCfg.name
  ]  else [
    builderGeneric
    "-c"
  ]
);
```

File: nixos/modules/system/boot/loader/raspberrypi/raspberrypi.nix [hunk 2]
```diff
@@ -4,13 +4,16 @@ with lib;
...
   cfg = config.boot.loader.raspberryPi;
+  dtCfg = config.hardware.deviceTree;
 
   builderUboot = import ./uboot-builder.nix { inherit pkgs configTxt; inherit (cfg) version; };
   builderGeneric = import ./raspberrypi-builder.nix { inherit pkgs configTxt; };
 
   builder =
     if cfg.uboot.enable then
-      "${builderUboot} -g ${toString cfg.uboot.configurationLimit} -t ${timeoutStr} -c"
+      "${builderUboot} -g ${toString cfg.uboot.configurationLimit} -t ${timeoutStr}"
```
Comment (klemensn) [inline nixos/modules/system/boot/loader/raspberrypi/raspberrypi.nix:14]: Why not just append the optional -n name after -c and thus diff churn?

**Review by klemensn — COMMENTED** [review]
Device-tree specifications and blobs are known to break on updates so being able to use your own (hopefully known-to-work) file is definitely useful.

**Comment by sorki** [comment]
I'm closing this since the `boot.loader.raspberryPi` was deprecated in https://github.com/NixOS/nixpkgs/pull/241534

You should migrate to `boot.loader.generic-extlinux-compatible` and you can find some pointers in https://github.com/NixOS/nixpkgs/pull/241534#issuecomment-1764896962 while we figure out a better upgrade path which is being worked on at https://github.com/NixOS/nixpkgs/pull/261857

**Pull request closed without merging**